### ISSUE TITLE
favicons not appearing after fetching process was completed

### ISSUE THREAD

**Issue by CrazieIvan** [aaFn/Bookmark-search-plus-2#222]
Hi, 
I wanted to see favicons in my bookmarks, so I unticked "Disable favicon fetching".  I let the fetching process complete (it shows 0 remaining favicons).  I was expecting that when the process is done, all the favicons would appear, but I'm still mostly seeing the generic favicon on most bookmarks.  Favicon appears only when I click on a bookmark to load a page.

Am I doing something wrong, or did I misunderstand that the fetching process should replace all the existing favicons?
Thank you for your time!

![2022-02-03_072702](https://user-images.githubusercontent.com/69615600/152287186-a0725c3a-6691-4137-925d-e756054f3ec4.jpg)

**Comment by aaFn** [comment]
Hello @CrazieIvan , correct, since several releases now, Firefox has been gradually increasing and enforcing protections, notably on CORS (Cross Origin Resources Sharing).

The result is that fetching the favicons in background by BSP2 is more and more receiving an error from FF, because BSP2 is seen as another origin as I get it.

This of course all depends on the CORS rules that web sites configure on their side, which FF is respecting more and more, or the ones they forgot to configure, which are now preventing such accesses.
Hence the symptom that less and less favicon fetches are succesful.

Now, accessing an open tab or loading a page are a second separate channel of providing favicons to BP2 = this is done by the page itself, so no CORS problem, and made available by FF API in the form of notification to BSP2, So that one always works, and catches up on favicons which fell in error during the background fetch.

I will leave this issue open to investigate once again if there is something else I can do in BSP2 background task, or if I can tweak a bit on the CORS complex subject to improve something, but no guarantee.

Last word: all this comes from the fact that contrary to Chrome, FF does not make the favicon DB available to Add-ons 😢 
This is https://bugzilla.mozilla.org/show_bug.cgi?id=1411120 and https://bugzilla.mozilla.org/show_bug.cgi?id=1315616 .. please vote on them !

**Comment by CrazieIvan** [comment]
That's unfortunate.  It seems FF made things very difficult or impossible.  
How do I vote for those issues?  I can't find the vote button.

**Comment by aaFn** [comment]
You need to have an id on bugzilla. Then log in, go to the Details section of the bug near top, and press the "Vote" button

**Comment by CrazieIvan** [comment]
aah it's hidden under Details, that's so strange.  OK thanks.

**Comment by chr7** [comment]
As I see, the problem is just the initial setup (=reading all favicons once). What about providing a small script (e.g. Python) to handle that initial setup? It could read the favicons from Firefox's sqlite DB and store them in some place from where BSP2 can import them.

**Comment by elandorr** [comment]
I was about to ask how to force a refresh of favicons before I saw this thread.

> This is https://bugzilla.mozilla.org/show_bug.cgi?id=1411120 and https://bugzilla.mozilla.org/show_bug.cgi?id=1315616 .. please vote on them !

There won't be proper favicon support then. Waiting for Mozzarella to fix basics means we might get a fix at our funeral. They're too busy adding new animations and telemetry and 'studies'. The whole reason you started this add-on in the first place is a literal decade old.

Thanks for your effort though!

> What about providing a small script (e.g. Python) to handle that initial setup?

If you write a tiny one I can read in 10 seconds I'd run it, maybe. Personally, I think that's just not worth the effort. People who don't have python would first need to get it and figure out how to run it on their own. An add-on can't trigger it. I do have python and all, but I don't think I could be bothered either. 

> Now, accessing an open tab or loading a page are a second separate channel of providing favicons to BP2 = this is done by the page itself, so no CORS problem, and made available by FF API in the form of notification to BSP2, So that one always works, and catches up on favicons which fell in error during the background fetch.

This works well and will get them all over time. Does this deal with changed favicons however?

**Comment by aaFn** [comment]
@elandorr , yes the method of getting favicons when tabs are open with corresponding bookmark URLs works with changed favicons. The only requirement is that there is no redirect on the URL, and so the tab URL is the one in the bookmark (else, redo the bookmark to point at the redirected-to page)

**Comment by CrazieIvan** [comment]
> @aaFn I personally don't care all that much about favicons, but it's nice that you're doing details like this too! I use BSP2 every day.
It's not my business how you use BSP2 and why you don't care about favicons.  I just want to explain why they are important and why they do serve a purpose for users like me who want to use BSP2 as the permanent bookmarks menu.  Favicons make bookmarks browsing MUCH faster and much easier on the eyes, because when I see a familiar icon, I don't have to read the bookmark name.  If all the favicons are missing, then recognizing bookmarks is much much slower, and strains the eyes because I have to read every bookmark name.  Imagine how you would feel if all your application icons on your desktop looked identical and you had to read all the names instead.  
If you don't care about favicons, that probably means you don't use BSP2 as the permanent bookmarks menu replacement.

**Comment by CrazieIvan** [comment]
I don't know why you got that impression, but there was nothing offensive about what you said.  
The only reason I responded with my take was because I think it's important to provide an alternative perspective about the importance of favicons.  That's all.
And most of my bookmarks are from different websites, so icons are pretty useful.

**Comment by elandorr** [comment]
All good, I see your point as well. Good luck with convincing FF! The current way of refreshing on next page load could be exploited maybe if addons can open windows on their own.

**Comment by aaFn** [comment]
Hello all, I am uploading to AMO version 2.0.101 where I added robustness and made some tuning on fetch() paramters to improve success rates, and I found a few remaining bugs which were also decreasing success rates for favicon fetching.

Hopefully with those improvements the success rates should be drastically going up again.
Let me know.

**Comment by CrazieIvan** [comment]
> All good, I see your point as well. Good luck with convincing FF! The current way of refreshing on next page load could be exploited maybe if addons can open windows on their own.

The problem with text is that it doesn't convey tone very well.  When I said it's none of my business, all I meant was that it's not my place to judge why you don't care.  :-)

@aaFn Is there something we need to do to restart fetching, or it wlll just do it once it's already active?

**Comment by aaFn** [comment]
To trigger all fetching again, do the following in the BSP2 options page:
- Disable favicon fetching (that will remove all favicons)
- Re-enable it (that will trigger fetch in background)

You can follow the progress in the _Statistics_ box.

Else, you can do, for each of the ones that you want to refresh:
- Right-click -> Advanced -> Refresh favicon
- Or click on it to load in a tab

.. but this is manual then ..

If needed, I could add an option to Trigger re-fetch of all favicons not successful so far.

**Comment by chr7** [comment]
Works great, thanks a lot!

**Comment by Gitoffthelawn** [comment]
I think it would be wonderful to have an option to trigger re-fetch of all favicons not successful so far.

BTW, thinking aloud here, in cases where the current API does not work, instead of specifically trying to access a favicon, would it be possible for BSP2 to simply try to download a couple of the filenames on the target server most common for favicons (e.g. `/favicon.ico`)?  Maybe it's what you are already doing... I haven't reviewed the code in a while.

And separately, who wants to sign a petition to make it a criminal offense to not include a favicon on your website? 🤣  It's 2022 people... c'mon, if it's worth creating a site, it's worth having a favicon! 😸

**Comment by aaFn** [comment]
@Gitoffthelawn , yes I already resort to fallback historical things like site-base-URL/favicon.ico in the code, when I cannot fetch through the discovered URLs .. that improves the success ratio by the way, so I confirm this is useful.

@elandorr, I summarized my action there https://github.com/aaFn/Bookmark-search-plus-2/wiki, mainly tuning of params related to fetch and cors after trial and errors and packet debug to verify, and a few bugs + robustness added for redirects.

On BSP2 rendering the huge list of of favicons lag-free (or nearly ..), I had to tune the code a lot, using clones, templates, background images (which I found to be faster), space reservation from start to avoid recalc / reflow .. etc .. I guess I am close to optimum I can do with FF and HTML/CSS in current state .. maybe it could be faster with webasm, but I am not familiar with that technology yet.

Ok, I'll look at the button to retry all unsuccessful favicons.
